moveit_core/trajectory_processing/src/cartesian_speed.cpp

```diff
@@ -0,0 +1,110 @@
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/* Author: Benjamin Scholz */
+
+#include <moveit/trajectory_processing/cartesian_speed.h>
+
+namespace trajectory_processing
+{
+bool setMaxCartesianEndEffectorSpeed(robot_trajectory::RobotTrajectory& trajectory, const double speed,
+                                     std::string end_effector)
+{
+  if (speed <= 0.0)
+  {
+    ROS_ERROR_STREAM_NAMED("trajectory_processing.cartesian_speed", "End effector speed must be greater than 0.");
+    return false;
+  }
+  if (end_effector.empty())
+  {
+    std::vector<std::string> tips;
+    trajectory.getGroup()->getEndEffectorTips(tips);
+    if (tips.empty())
+    {
+      ROS_ERROR_STREAM_NAMED("trajectory_processing.cartesian_speed",
+                             "No end effector defined for group attached to trajectory.");
+      return false;
+    }
+    end_effector = tips[0];
+  }
+
+  size_t num_waypoints = trajectory.getWayPointCount();
+  if (num_waypoints == 0)
+    return false;
+
+  robot_state::RobotStatePtr kinematic_state = trajectory.getFirstWayPointPtr();
+
+  // do forward kinematics to get cartesian positions of end effector for current waypoint
+  Eigen::Isometry3d current_end_effector_state = kinematic_state->getGlobalLinkTransform(end_effector);
+  Eigen::Isometry3d next_end_effector_state;
+  double euclidean_distance, new_time_diff, old_time_diff;
+  std::vector<double> time_diff(num_waypoints - 1, 0.0);
+
+  for (size_t i = 0; i < num_waypoints - 1; i++)
+  {
+    // get state for next waypoint
+    kinematic_state = trajectory.getWayPointPtr(i + 1);
+
+    // do forward kinematics to get cartesian positions of end effector for next waypoint
+    next_end_effector_state = kinematic_state->getGlobalLinkTransform(end_effector);
+
+    // get euclidean distance between the two waypoints
+    euclidean_distance =
+        pow(pow(next_end_effector_state.translation()[0] - current_end_effector_state.translation()[0], 2) +
+                pow(next_end_effector_state.translation()[1] - current_end_effector_state.translation()[1], 2) +
+                pow(next_end_effector_state.translation()[2] - current_end_effector_state.translation()[2], 2),
+            0.5);
+
+    new_time_diff = (euclidean_distance / speed);
+    old_time_diff = trajectory.getWayPointDurationFromPrevious(i + 1);
+
+    // if constraints allow save the new time difference between waypoints
+    if (new_time_diff > old_time_diff)
+    {
+      time_diff[i] = new_time_diff;
+    }
+    else
+    {
+      ROS_WARN_ONCE_NAMED("trajectory_processing.cartesian_speed",
+                          "Desired cartesian end-effector speed is not reached because of joint velocity constraints.");
+      time_diff[i] = old_time_diff;
+    }
+    // update the current_end_effector_state for next iteration
+    current_end_effector_state = next_end_effector_state;
+  }
```

Comment:
I'd suggest refactoring to keep more variables within the local scope of the `for` loop. The loop is basically iterating over pairs of waypoint indices, and since the state transforms depend on new values of `kinematic_state` that are gotten for each waypoint index, I think it would be cleaner to just determine new values of current and next based on the index i for the current waypoint.

Something like this could be one solution:
```suggestion
  double new_time_diff, old_time_diff;
  std::vector<double> time_diff(num_waypoints - 1, 0.0);

  for (size_t i = 0; i < num_waypoints - 1; i++)
  {
    Eigen::Isometry3d current_end_effector_state = trajectory.getWayPointPtr(i)->getGlobalLinkTransform(end_effector); 
    Eigen::Isometry3d next_end_effector_state = trajectory.getWayPointPtr(i + 1)->getGlobalLinkTransform(end_effector); 

    // get euclidean distance between the two waypoints
    double euclidean_distance =
        pow(pow(next_end_effector_state.translation()[0] - current_end_effector_state.translation()[0], 2) +
                pow(next_end_effector_state.translation()[1] - current_end_effector_state.translation()[1], 2) +
                pow(next_end_effector_state.translation()[2] - current_end_effector_state.translation()[2], 2),
            0.5);

    new_time_diff = (euclidean_distance / speed);
    old_time_diff = trajectory.getWayPointDurationFromPrevious(i + 1);

    // if constraints allow save the new time difference between waypoints
    if (new_time_diff > old_time_diff)
    {
      time_diff[i] = new_time_diff;
    }
    else
    {
      ROS_WARN_ONCE_NAMED("trajectory_processing.cartesian_speed",
                          "Desired cartesian end-effector speed is not reached because of joint velocity constraints.");
      time_diff[i] = old_time_diff;
    }
  }
```